nexus/src/db/datastore.rs

```diff
@@ -2050,29 +2131,30 @@ impl DataStore {
 
     pub async fn project_update_vpc(
         &self,
-        vpc_id: &Uuid,
+        opctx: &OpContext,
+        authz_vpc: &authz::Vpc,
         updates: VpcUpdate,
-    ) -> Result<(), Error> {
-        use db::schema::vpc::dsl;
+    ) -> UpdateResult<Vpc> {
+        opctx.authorize(authz::Action::Modify, authz_vpc).await?;
 
+        use db::schema::vpc::dsl;
         diesel::update(dsl::vpc)
             .filter(dsl::time_deleted.is_null())
-            .filter(dsl::id.eq(*vpc_id))
+            .filter(dsl::id.eq(authz_vpc.id()))
             .set(updates)
-            .execute_async(self.pool())
+            .returning(Vpc::as_returning())
+            .get_result_async(self.pool_authorized(opctx).await?)
             .await
             .map_err(|e| {
                 public_error_from_diesel_pool(
                     e,
-                    ErrorHandler::NotFoundByLookup(
-                        ResourceType::Vpc,
-                        LookupType::ById(*vpc_id),
-                    ),
+                    ErrorHandler::NotFoundByResource(authz_vpc),
                 )
-            })?;
-        Ok(())
+            })
     }
 
+    // TODO-security TODO-cleanup Remove this function.  Update callers to use
+    // vpc_lookup_by_path() or vpc_fetch() instead.
```

Comment:
It looks like there are two call sites for the `Nexus::vpc_fetch_by_name` method. Would it be a good idea to just tackle those here, so we don't lose track?

Reply:
I looked at that, but one of the callers is `instance_create_network_interface`, which needs the actual `Vpc`.  To use `vpc_fetch()`, it would need an `OpContext` so it can do an authz check.  It seemed like that would involve adding authn/authz to more endpoints and I'm trying to keep these PRs as small as reasonable.

That said, I just (re)discovered that `instance_create_network_interface` is not used at all, so maybe this is worth looking at now.

Reply:
`Nexus::instance_create_network_interface` is used during the current instance creation saga, I think. I added auth to a new version of that method as part of #713. One of your comments there, though, was about the fact that the auth checks it performs are just on the instance. It should probably add or change that to be on the VPC or VPC Subnet, with the `CreateChild` permissions. All that is to say, we should leave this for this PR. I can either base #713 on this and remove the call to `vpc_fecth_by_name`, or I can do that as part of a follow-up PR.

Reply:
I believe it's `DataStore::instance_create_network_interface` that's used in the saga.  `Nexus::instance_create_network_interface` appears to be unused, at least according to rust-analyzer.

Given that, I looked at fixing this TODO in this PR, but one of the other callers of `vpc_fetch_by_name` is `Nexus::project_lookup_vpc`.  This is the one we discussed in the other thread.  This one is hard to remove or fix without fixing a bunch more endpoints.  So I think I will go ahead and defer this.

Reply:
> I believe it's DataStore::instance_create_network_interface

Indeed it is, my bad!

>  So I think I will go ahead and defer this.

Sounds good to me!